navigation/pose_action_server/src/pose_action_server_ros.cpp

```diff
@@ -0,0 +1,148 @@
+    const std::shared_ptr<GoalHandleFilteredPose> goal_handle) {
+    RCLCPP_INFO(this->get_logger(), "Goal accepted. Processing poses.");
+    is_executing_action_ = true;
+    active_goal_handle_ = goal_handle;
+}
+
+void PoseActionServerNode::pose_callback(
+    const geometry_msgs::msg::PoseStamped::ConstSharedPtr pose_msg) {
+    if (!is_executing_action_)
+        return;
+
+    pose_queue_.push_back(*pose_msg);
+
+    geometry_msgs::msg::PoseStamped feedback_msg = *pose_msg;
+    auto feedback =
+        std::make_shared<vortex_msgs::action::FilteredPose::Feedback>();
+    feedback->current_pose = feedback_msg;
+    active_goal_handle_->publish_feedback(feedback);
+
+    if (pose_queue_.size() == num_measurements_) {
+        geometry_msgs::msg::PoseStamped filtered_pose;
+        filtered_pose.header = pose_msg->header;
+
+        std::vector<double> x_positions;
+        std::vector<double> y_positions;
+        std::vector<double> z_positions;
+        std::vector<double> yaw_angles;
+
+        for (const auto& pose : pose_queue_) {
+            x_positions.push_back(pose.pose.position.x);
+            y_positions.push_back(pose.pose.position.y);
+            z_positions.push_back(pose.pose.position.z);
+
+            double qz = pose.pose.orientation.z;
+            double qw = pose.pose.orientation.w;
+            double yaw = 2.0 * std::atan2(qz, qw);
+            yaw_angles.push_back(yaw);
+        }
+
+        std::sort(x_positions.begin(), x_positions.end());
+        std::sort(y_positions.begin(), y_positions.end());
+        std::sort(z_positions.begin(), z_positions.end());
+
+        size_t mid = num_measurements_ / 2;
+
+        if (num_measurements_ % 2 == 0) {
+            filtered_pose.pose.position.x =
+                (x_positions[mid - 1] + x_positions[mid]) / 2.0;
+            filtered_pose.pose.position.y =
+                (y_positions[mid - 1] + y_positions[mid]) / 2.0;
+            filtered_pose.pose.position.z =
+                (z_positions[mid - 1] + z_positions[mid]) / 2.0;
+        } else {
+            filtered_pose.pose.position.x = x_positions[mid];
+            filtered_pose.pose.position.y = y_positions[mid];
+            filtered_pose.pose.position.z = z_positions[mid];
+        }
+
+        std::vector<double> unwrapped_yaw = yaw_angles;
+        for (size_t i = 1; i < unwrapped_yaw.size(); ++i) {
+            double diff = unwrapped_yaw[i] - unwrapped_yaw[i - 1];
+            if (diff > M_PI) {
+                unwrapped_yaw[i] -= 2.0 * M_PI;
+            } else if (diff < -M_PI) {
+                unwrapped_yaw[i] += 2.0 * M_PI;
+            }
+        }
+
+        std::sort(unwrapped_yaw.begin(), unwrapped_yaw.end());
+        double median_yaw = unwrapped_yaw[mid];
+        if (num_measurements_ % 2 == 0) {
+            median_yaw = (unwrapped_yaw[mid - 1] + unwrapped_yaw[mid]) / 2.0;
+        } else {
+            median_yaw = unwrapped_yaw[mid];
+        }
```

Comment:
Whenever you deal with raw euler-representation you *will* have discontinuities somewhere, so I can't look at this code and immediately say "yes this will work 100% for all possible inputs" -> That to me signals the need for a refactor.

There's no need to split up larger functions into smaller ones "just because". However, it let's you define names for said functions, which will greatly help with readability in cases like these.